src/counters.ts

```diff
@@ -0,0 +1,694 @@
+
+	}
+
+	private async stopCollecting(): Promise<void> {
+		await this.countersCollectionPromise;
+		clearInterval(this.countersTimer);
+		if (this.countersTimer) {
+			this.countersTimer.unref();
+			this.countersTimer = null;
+		}
+		await this.countersCollectionPromise;
+	}
+
+	private async startPopulatingProcessInfos(): Promise<void> {
+		assert(this.processesToTrackTimer === null, `the processesToTrackTimer should be null when we startPopulatingProcessInfos`);
+		const getProcessInfos = async () => {
+
+			this.processesToTrack = await getChildrenTree(this.pid, this.includeParent);
+		};
+		// kickoff the collection of processes to track and await 
+		// them to be collected for the first time so that processesToTrack is initialized
+		this.processesToTrackUpdationPromise = getProcessInfos();
+		await this.processesToTrackUpdationPromise;
+		// set a timer to keep getting processes to track at regular intervals.
+		this.processesToTrackTimer = setInterval(() => {
+			// start updating processesToTrack unless we are already doing one.
+			if (!this.processesToTrackUpdationInProgress) {
+				this.processesToTrackUpdationInProgress = true;
+				this.processesToTrackUpdationPromise = getProcessInfos();
+				this.processesToTrackUpdationInProgress = false;
+			}
+		}, Counters.ProcessInfoUpdationInterval);
+	}
+
+	private async startCollecting(): Promise<void> {
+		assert(this.countersTimer === null, `the countersTimer should be null when we startCollecting`);
+		const collectCounters = async () => {
+			const cs: ProcessStatistics = {
+				cpu: (osu.cpu.average()).avgTotal,
+				memory: os.totalmem() - os.freemem(),
+				ppid: undefined,
+				pid: 0,
+				elapsed: os.uptime() * 1000,
+				timestamp: Date.now()
+			};
+			[cs, ...await this.getCounters()].forEach(processStatistics => {
+				const trace = require('debug')(`${logPrefix}:collectCounters:trace`);
+				Object.keys(processStatistics).filter(key => (key !== "pid" && key !== "ppid")).forEach(prop => {
+					if (!this.collection[processStatistics.pid]) {
+						this.collection[processStatistics.pid] = {
+							pid: processStatistics.pid,
+							ppid: processStatistics.ppid
+						};
+					}
+					if (!this.collection[processStatistics.pid][prop]) {
+						this.collection[processStatistics.pid][prop] = [];
+					}
+					trace(`prop: ${prop}, propColl: ${jsonDump(this.collection[processStatistics.pid][prop])}`);
+					this.collection[processStatistics.pid][prop].push(processStatistics[prop]);
+				});
+			});
+		};
+		// Kickoff the first collection
+		this.countersCollectionPromise = collectCounters();
+		// Set a timer for future collection(s)
+		this.countersTimer = setInterval(() => {
+			// start collecting counters unless we are already collecting one set.
+			if (!this.countersCollectionInProgress) {
+				this.countersCollectionInProgress = true;
+				this.countersCollectionPromise = collectCounters();
+				this.countersCollectionInProgress = false;
+			}

```

Comment:
What's the purpose of this flag? Since you're not await'ing here it seems unlikely that you'll hit this multiple times

Reply:
This is because promises currently have no way of query the status or a pending promise, whether it is in progress, fulfilled or errored. So these flags are used to track the state of the corresponding promise.

In the following lines:

		this.countersTimer = setInterval(() => {
			if (!this.countersCollectionInProgress) {
				this.countersCollectionPromise = collectCounters();
			}
		}, this.collectionIntervalMs);

we use this flag to ensure that we do not schedule a new collection if the previous one has not finished.